nifi-nar-bundles/nifi-social-media-bundle/nifi-twitter-processors/src/main/java/org/apache/nifi/processors/twitter/TweetStreamService.java

```diff
@@ -0,0 +1,222 @@
+        this.endpoint = endpoint;
+        executorService.execute(new TweetStreamStarter());
+    }
+
+    /**
+     * This method would be called when we would like the stream to get stopped. The stream will be closed and the
+     * executorService will be shut down.
+     */
+    public void stop() {
+        if (stream != null) {
+            try {
+                stream.close();
+            } catch (IOException e) {
+                logger.error("Closing response stream failed", e);
+            }
+        }
+
+        executorService.shutdownNow();
+    }
+
+    private Long calculateBackoffDelay() {
+        long backoff = backoffMultiplier * backoffTime;
+        long delay = (backoff > maximumBackoff) ? maximumBackoff : backoff;
+        return delay;
+    }
+
+    private void scheduleStartStreamWithBackoff() {
+        // use exponential(by factor of 2) backoff in scheduling the next TweetStreamStarter
+        if (attemptCounter >= backoffAttempts) {
+            throw new ProcessException(String.format("Connection failed after maximum attempts [%d]", attemptCounter));
+        }
+        attemptCounter += 1;
+        long delay = calculateBackoffDelay();
+        backoffMultiplier *= 2;
+        logger.info("Scheduling new stream connection after delay [{} s]", delay);
+        executorService.schedule(new TweetStreamStarter(), delay, TimeUnit.SECONDS);
+    }
+
+    private void resetBackoff() {
+        attemptCounter = 0;
+        backoffMultiplier = 1L;
+    }
+
+    private class TweetStreamStarter implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (endpoint.equals(StreamEndpoint.SAMPLE_ENDPOINT)) {
+                    stream = api.tweets().sampleStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                } else {
+                    stream = api.tweets().searchStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                }
+                executorService.execute(new TweetStreamHandler());
+            } catch (ApiException e) {
+                stream = null;
+                logger.warn("Twitter Stream [{}] API connection failed: HTTP {}", endpoint.getEndpointName(), e.getCode(), e);
+                scheduleStartStreamWithBackoff();
+            } catch (Exception e) {
+                stream = null;
+                logger.warn("Twitter Stream [{}] connection failed", endpoint.getEndpointName(), e);
+                scheduleStartStreamWithBackoff();
+            }
+        }
+    }
+
+    private class TweetStreamHandler implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream))) {
+                String tweetRecord = reader.readLine();
+                while (tweetRecord != null) {
+                    queue.offer(tweetRecord);
```

Comment:
The `offer()` method can return `false`, indicating that the queue is full, which result in missing records. The `put()` method blocks, which seems like the best behavior for now, in order to avoid reading more from the stream until space is available in the queue. The `put()` method can throw an `InterruptedException`, signalling that processing is being terminated. Recommend adding a second catch block for that condition, and logging that adding Tweet to the queue failed.